File: backend/src/auth/auth.service.ts
import { Injectable, UnauthorizedException, ConflictException } from '@nestjs/common';
import { JwtService } from '@nestjs/jwt';
import * as bcrypt from 'bcrypt';
import { PrismaService } from '../prisma/prisma.service';
import { RegisterDto, LoginDto } from './dto/auth.dto';

@Injectable()
export class AuthService {
constructor(private prisma: PrismaService, private jwtService: JwtService) {}

  async register(dto: RegisterDto) {
    const existing = await this.prisma.user.findUnique({ where: { email: dto.email } });
    if (existing) throw new ConflictException('Email already exists');
    const hash = await bcrypt.hash(dto.password, 10);
    const user = await this.prisma.user.create({
      data: {
        email: dto.email,
        password_hash: hash,
        full_name: dto.full_name,
        avatar_url: dto.avatar_url,
        bio: dto.bio,
      },
    });
    const token = this.jwtService.sign({ sub: user.id, email: user.email, role: user.role });
    const { password_hash, ...userWithoutPwd } = user;
    return { user: userWithoutPwd, token };
  }

  async login(dto: LoginDto) {
    const user = await this.prisma.user.findUnique({ where: { email: dto.email } });
    if (!user) throw new UnauthorizedException('Invalid credentials');
    const valid = await bcrypt.compare(dto.password, user.password_hash);
    if (!valid) throw new UnauthorizedException('Invalid credentials');
    const token = this.jwtService.sign({ sub: user.id, email: user.email, role: user.role });
    const { password_hash, ...userWithoutPwd } = user;
    return { user: userWithoutPwd, token };
  }

  async forgotPassword(email: string) {
    // In production, send email with reset link. For now just return message.
    return { message: 'If email exists, reset link sent' };
  }
}

File: frontend/lib/services/auth_service.dart
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import 'api_service.dart';

class AuthService {
  final ApiService _api = ApiService();
  final FlutterSecureStorage _storage = const FlutterSecureStorage();

  Future<Map<String, dynamic>?> login(String email, String password) async {
    try {
      final res = await _api.post('/auth/login', {'email': email, 'password': password});
      if (res.statusCode == 200) {
        await _storage.write(key: 'access_token', value: res.data['token']);
        return res.data['user'];
      }
    } catch (e) {
      print('Login error: $e');
    }
    return null;
  }

  Future<Map<String, dynamic>?> register(String email, String password, String fullName, {String? avatarUrl, String? bio}) async {
    try {
      final res = await _api.post('/auth/register', {
        'email': email,
        'password': password,
        'full_name': fullName,
        'avatar_url': avatarUrl,
        'bio': bio,
      });
      if (res.statusCode == 201) {
        await _storage.write(key: 'access_token', value: res.data['token']);
        return res.data['user'];
      }
    } catch (e) {
      print('Register error: $e');
    }
    return null;
  }

  Future<void> logout() async {
    await _storage.delete(key: 'access_token');
  }

  Future<String?> getToken() async {
    return await _storage.read(key: 'access_token');
  }
}